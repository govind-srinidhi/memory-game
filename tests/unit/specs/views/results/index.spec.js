/**
 * @file It contains unit tests for results overview page.
 */
import { mount, createLocalVue } from "@vue/test-utils"
import Vuetify from "vuetify"
import Vuex from "vuex"
import VueRouter from "vue-router"
import ResultsPage from "@/views/results"

let wrapper

const results = [{
  value: 23,
  date: new Date("2021-04-11T19:30:00"),
  valid: "Incorrect"
}, {
  value: 1,
  date: new Date("2021-04-11T19:30:01"),
  valid: "Incorrect"
}, {
  value: 44,
  date: new Date("2021-04-11T19:30:02"),
  valid: "Correct"
}, {
  value: 45,
  date: new Date("2021-04-11T19:30:03"),
  valid: "Correct"
}]

const state = {
  userAnswers: results
}

const getters = {
  "results/userAnswers": state => state.userAnswers,
}

const mutations = {
  setUserAnswers: (state, userAnswers) => {
    state.userAnswers = userAnswers
  }
}

const router = new VueRouter()
let store

const mountResultsPage = () => {
  const localVue = createLocalVue()

  localVue.prototype.$TABLES = {
    RESULTS: {
      headers: [{
        text: "Value",
        align: "left",
        value: "value",
        class: "subheading primary lighten-2 white--text subtitle-2 font-weight-bold"
      }, {
        text: "Time",
        align: "left",
        value: "selectedTime",
        class: "subheading primary lighten-2 white--text subtitle-2 font-weight-bold"
      }, {
        text: "Valid",
        align: "left",
        value: "valid",
        class: "subheading primary lighten-2 white--text subtitle-2 font-weight-bold"
      }],
      noDataText: "No data found.",
      itemsPerPage: 10,
      footer: {
        itemsPerPageOptions: [10, 20, 30],
        itemsPerPageText: "Items per page",
        showFirstLastPage: true,
        showCurrentPage: true
      }
    }
  }
  localVue.prototype.$CONSTANTS = {
    GAME_RESULTS: {
      CORRECT: "Correct",
      INCORRECT: "Incorrect"
    },
    LABELS: {
      PLAY_AGAIN: "Play again"
    }
  }
  localVue.use(Vuex)
  localVue.use(Vuetify)
  localVue.use(VueRouter)

  const vuetify = new Vuetify()

  store = new Vuex.Store({
    state,
    getters,
    mutations
  })

  wrapper = mount(ResultsPage, {
    sync: false,
    localVue,
    store,
    router,
    vuetify
  })
}

const resultsTable = () => wrapper.findComponent({ ref: "table_results" })
const playAgain = () => wrapper.findComponent({ ref: "button_play_again" })

describe("results page", () => {
  beforeEach(() => {
    mountResultsPage()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it("renders the results page", () => {
    expect(wrapper).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("compute all computed properties correctly", () => {
    expect(wrapper.vm.resultsForTable).toEqual([{
      value: 23,
      selectedTime: new Date("2021-04-11T19:30:00").toLocaleString(),
      valid: "Incorrect"
    }, {
      value: 1,
      selectedTime: new Date("2021-04-11T19:30:01").toLocaleString(),
      valid: "Incorrect"
    }, {
      value: 44,
      selectedTime: new Date("2021-04-11T19:30:02").toLocaleString(),
      valid: "Correct"
    }, {
      value: 45,
      selectedTime: new Date("2021-04-11T19:30:03").toLocaleString(),
      valid: "Correct"
    }])

    expect(wrapper.vm.headersForTable).toEqual([{
      align: "left",
      class: "subheading primary lighten-2 white--text subtitle-2 font-weight-bold",
      text: "Value",
      value: "value"
    }, {
      align: "left",
      class: "subheading primary lighten-2 white--text subtitle-2 font-weight-bold",
      text: "Time",
      value: "selectedTime"
    }, {
      align: "left",
      class: "subheading primary lighten-2 white--text subtitle-2 font-weight-bold",
      text: "Valid",
      value: "valid"
    }])
    expect(wrapper.vm.footersForTable).toEqual({
      itemsPerPageOptions: [10, 20, 30],
      itemsPerPageText: "Items per page",
      showCurrentPage: true,
      showFirstLastPage: true
    })
    expect(wrapper.vm.resultMessage).toEqual("You've answered 50% of questions correctly! Review your response")

    store.commit("setUserAnswers", [{
      value: 1,
      date: Date("2021-04-11T19:30:00"),
      valid: "Correct"
    }, {
      value: 23,
      date: Date("2021-04-11T19:30:01"),
      valid: "Correct"
    }, {
      value: 44,
      date: Date("2021-04-11T19:30:02"),
      valid: "Correct"
    }, {
      value: 45,
      date: Date("2021-04-11T19:30:03"),
      valid: "Correct"
    }, {
      value: 46,
      date: Date("2021-04-11T19:30:02"),
      valid: "Correct"
    }, {
      value: 47,
      date: Date("2021-04-11T19:30:03"),
      valid: "Correct"
    }, {
      value: 50,
      date: Date("2021-04-11T19:30:03"),
      valid: "Incorrect"
    }, {
      value: 48,
      date: Date("2021-04-11T19:30:03"),
      valid: "Incorrect"
    }])
    expect(wrapper.vm.resultMessage).toEqual("Good attempt. You've answered 75% of questions correctly! Review your response")

    store.commit("setUserAnswers", [{
      value: 1,
      date: Date("2021-04-11T19:30:00"),
      valid: "Correct"
    }, {
      value: 23,
      date: Date("2021-04-11T19:30:01"),
      valid: "Correct"
    }, {
      value: 44,
      date: Date("2021-04-11T19:30:02"),
      valid: "Correct"
    }, {
      value: 45,
      date: Date("2021-04-11T19:30:03"),
      valid: "Correct"
    }])
    expect(wrapper.vm.resultMessage).toEqual("Congratulations. You've won the game!")
  })

  it("renders results table correctly", async () => {
    expect(resultsTable().exists()).toBeTruthy()

    const tableHeaders = resultsTable().findAll("thead tr th")

    expect(tableHeaders.at(0).exists()).toBeTruthy()
    expect(tableHeaders.at(0).text()).toContain("Value")
    expect(tableHeaders.at(1).exists()).toBeTruthy()
    expect(tableHeaders.at(1).text()).toContain("Time")
    expect(tableHeaders.at(2).exists()).toBeTruthy()
    expect(tableHeaders.at(2).text()).toContain("Valid")

    const tableRows = resultsTable().findAll("tbody tr")

    for (const iterator in results) {
      const row = tableRows.at(iterator)

      expect(row.exists()).toBeTruthy()

      const cells = row.findAll("td")

      expect(cells.at(0).exists()).toBeTruthy()
      expect(cells.at(0).text()).toContain(wrapper.vm.resultsForTable[iterator].value)
      expect(cells.at(1).exists()).toBeTruthy()
      expect(cells.at(1).text()).toContain(wrapper.vm.resultsForTable[iterator].selectedTime)
      expect(cells.at(2).exists()).toBeTruthy()
      expect(cells.at(2).text()).toContain(wrapper.vm.resultsForTable[iterator].valid)
    }
  })

  it("navigates to home page, on clicking play again button", async () => {
    VueRouter.prototype.replace = jest.fn()
    playAgain().trigger("click")
    await wrapper.vm.$nextTick()

    expect(router.replace).toBeCalledWith("/")
  })
})