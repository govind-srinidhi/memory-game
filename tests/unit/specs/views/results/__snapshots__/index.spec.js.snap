// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`results page renders the results page 1`] = `
<div class="container">
  <div class="d-block pa-4 subtitle-1 text-center">You've answered 50% of questions correctly! Review your response</div>
  <div class="v-data-table elevation-1 v-data-table--fixed-header v-data-table--has-bottom theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Value: Not sorted. Activate to sort ascending." aria-sort="none" class="text-left subheading primary lighten-2 white--text subtitle-2 font-weight-bold sortable"><span>Value</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Time: Not sorted. Activate to sort ascending." aria-sort="none" class="text-left subheading primary lighten-2 white--text subtitle-2 font-weight-bold sortable"><span>Time</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Valid: Not sorted. Activate to sort ascending." aria-sort="none" class="text-left subheading primary lighten-2 white--text subtitle-2 font-weight-bold sortable"><span>Valid</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-left">23</td>
            <td class="text-left">4/11/2021, 7:30:00 PM</td>
            <td class="text-left">Incorrect</td>
          </tr>
          <tr class="">
            <td class="text-left">1</td>
            <td class="text-left">4/11/2021, 7:30:01 PM</td>
            <td class="text-left">Incorrect</td>
          </tr>
          <tr class="">
            <td class="text-left">44</td>
            <td class="text-left">4/11/2021, 7:30:02 PM</td>
            <td class="text-left">Correct</td>
          </tr>
          <tr class="">
            <td class="text-left">45</td>
            <td class="text-left">4/11/2021, 7:30:03 PM</td>
            <td class="text-left">Correct</td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Items per page<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-12" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Items per page" id="input-12" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">1-4 of 4</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="First page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-page-first theme--light"></i></span></button><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div><span>1</span>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Last page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-page-last theme--light"></i></span></button></div>
    </div>
  </div><button type="button" class="mt-4 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary lighten-2" text-color="black"><span class="v-btn__content">Play again</span></button>
</div>
`;
